chain-core/src/common/merkle_tree.rs

```diff
@@ -81,92 +81,97 @@ impl<T> Tree<T> {
             } => {
                 if &hash(value, NodeType::Leaf) == node_hash {
                     Some(Path {
-                        node_hash: *node_hash,
-                        sibling: None,
-                        sub_path: None,
+                        leaf_hash: node_hash.clone(),
+                        nodes: vec![],
                     })
                 } else {
                     None
                 }
             }
-            Tree::Node { hash, left, right } => match left.generate_path(value.clone()) {
-                None => right.generate_path(value).map(|mut path| {
-                    path.sibling = Some(Sibling::Left(left.hash()));
-
-                    Path {
-                        node_hash: *hash,
-                        sibling: None,
-                        sub_path: Some(Box::new(path)),
-                    }
-                }),
-                Some(mut path) => {
-                    path.sibling = Some(Sibling::Right(right.hash()));
-
-                    Some(Path {
-                        node_hash: *hash,
-                        sibling: None,
-                        sub_path: Some(Box::new(path)),
+            Tree::Node {
+                hash: node_hash,
+                left,
+                right,
+            } => left.generate_path(value).map_or_else(
+                || {
+                    right.generate_path(value).map(|mut path| {
+                        path.nodes.push(PathNode {
+                            node_hash: node_hash.clone(),
+                            child_hash: left.hash(),
+                            hash_side: Side::Left,
+                        });
+                        path
                     })
-                }
-            },
+                },
+                |mut path| {
+                    path.nodes.push(PathNode {
+                        node_hash: node_hash.clone(),
+                        child_hash: right.hash(),
+                        hash_side: Side::Right,
+                    });
+                    Some(path)
+                },
+            ),
         }
     }
 }
 
-/// Sibling's hash
+/// Merkle path for inclusion proof
+
 #[derive(Debug, PartialEq, Eq, Clone, Encode, Decode)]
-pub enum Sibling {
-    Left(H256),
-    Right(H256),
+pub enum Side {
+    Left,
+    Right,
 }
 
-/// Merkle path for inclusion proof
 #[derive(Debug, PartialEq, Eq, Clone, Encode, Decode)]
-pub struct Path {
+pub struct PathNode {
     node_hash: H256,
-    sibling: Option<Sibling>,
-    sub_path: Option<Box<Path>>,
+    child_hash: H256, // hash of left child unless reversed
+    hash_side: Side,  // the side of hash child
```

Comment:
could add some explanation , which "side" means?

Reply:
Just added some ascii art to explain it: https://github.com/crypto-com/chain/pull/558/commits/ab63bbc327de13e451a8e43b300a6cb158e77706